internal/datastores/vsan_datastore_subresource.go

```diff
@@ -40,6 +40,11 @@ func VsanDatastoreSchema() *schema.Resource {
 				Optional:    true,
 				Description: "Enable vSAN deduplication and compression",
 			},
+			"esa_enabled": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Description: "Enable vSAN ESA",
+			},
```

Comment:
Wouldn't this conflict with `dedup_and_compression_enabled`?

```suggestion
			"esa_enabled": {
				Type:        schema.TypeBool,
				Optional:    true,
				Description: "Enable vSAN ESA",
				ConflictsWith: []string{"dedup_and_compression_enabled"},
			},
```

Reply:
Keep in mind that the following might also be required:

```go
			"dedup_and_compression_enabled": {
				Type:          schema.TypeBool,
				Optional:      true,
				Description:   "Enable vSAN deduplication and compression",
                ConflictsWith: []string{"esa_enabled"},
			},
```

Also, needs a `go fmt` since the alignment if off it seems.

Reply:
What might be the impact if this is not enabled first and OSA is use?

I don't think you can have it be `false` and then try to set it to [true](url) later on to get to ESA. I think that would need to fail. Same for the reverse.

I could be wrong but we should verify so that this is not confusing, and ensure that we add a test case and docs to ensure it's documented with examples.

Reply:
Hey @tenthirtyam, it's true that `esa_enabled` and `dedup_and_compression_enabled` cannot coexist.

The reason why I didn't add `ConflictsWith` is that it requires an absolute path to the conflicting property and that doesn't go well with list attributes. I'm pasting an excerpt from the docs for easier reference

`
	// Only absolute attribute paths, ones starting with top level attribute
	// names, are supported. Attribute paths cannot be accurately declared
	// for TypeList (if MaxItems is greater than 1), TypeMap, or TypeSet
	// attributes. To reference an attribute under a single configuration block
	// (TypeList with Elem of *Resource and MaxItems of 1), the syntax is
	// "parent_block_name.0.child_attribute_name".
	ConflictsWith []string
`

By the way, the API returns a neat error in this case and we should be okay in terms of UX

Reply:
As for your last comment - AFAIK you can't go back and forth between ESA and non-ESA.
Once you enable vSAN in ESA mode you need to tear down and rebuild your storage if you want to change it.

I'm not adding any new tests since ESA imposes requirements on the underlying ESX hosts and such tests would be difficult to run.
I used the existing tests (with minor modifications of course) to verify the change. I ran them on ESA-compatible hosts

Reply:
@tenthirtyam I'm closing this thread.
I've outlined the conflicts in the docs